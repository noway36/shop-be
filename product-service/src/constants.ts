export enum HttpCode {
    Ok = 200,
    NotFound = 404,
    InternalServerError = 500,
}
