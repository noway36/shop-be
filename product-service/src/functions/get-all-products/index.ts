import { handlerPath } from '@libs/handlerResolver';

export default {
    handler: `${handlerPath(__dirname)}/handler.main`,
    events: [
        {
            http: {
                method: 'get',
                path: 'products',
                cors: true,
                documentation: {
                    summary: 'get all available products',
                    description: 'get all available products',
                    methodResponses: [
                        {
                            statusCode: 200,
                            responseBody: {
                                description: 'all available products',
                            },
                            responseModels: {
                                'application/json': 'ProductItemsArray',
                            },
                        },
                        {
                            statusCode: 500,
                            responseBody: {
                                description: 'all available products',
                            },
                            responseModels: {
                                'application/json': 'object',
                            },
                        },
                    ],
                },
            },
        },
    ],
};
