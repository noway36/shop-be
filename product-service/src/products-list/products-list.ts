export interface IProduct {
    count: number;
    description: string;
    id: string;
    price: number;
    title: string;
}

export const PRODUCT_BY_ID_MAP = new Map<string, IProduct>([
    [
        '9a14812c-07e1-463e-9fa1-1a5179f48e61',
        {
            count: 4,
            description: 'Nice Product 1',
            id: '9a14812c-07e1-463e-9fa1-1a5179f48e61',
            price: 2,
            title: 'Nice Product 1',
        },
    ],
    [
        '164032bb-6512-4d5b-aae8-313fc365f012',
        {
            count: 6,
            description: 'Nice Product 2',
            id: '164032bb-6512-4d5b-aae8-313fc365f012',
            price: 10,
            title: 'Nice Product 2',
        },
    ],
    [
        'b3595014-634c-4b47-8250-cce3867d4919',
        {
            count: 7,
            description: 'Nice Product 3',
            id: 'b3595014-634c-4b47-8250-cce3867d4919',
            price: 23,
            title: 'Nice Product 3',
        },
    ],
    [
        '504e5779-c1bc-4498-a792-3b12839b4a7d',
        {
            count: 12,
            description: 'Nice Product 4',
            id: '504e5779-c1bc-4498-a792-3b12839b4a7d',
            price: 15,
            title: 'Nice Product 4',
        },
    ],

    [
        'e7a1919b-0bb8-4511-980b-553a332be219',
        {
            count: 7,
            description: 'Nice Product 5',
            id: 'e7a1919b-0bb8-4511-980b-553a332be219',
            price: 23,
            title: 'Nice Product 5',
        },
    ],
    [
        '00e95255-4243-426c-93e7-b9504a8bff89',
        {
            count: 8,
            description: 'Nice Product 6',
            id: '00e95255-4243-426c-93e7-b9504a8bff89',
            price: 15,
            title: 'Nice Product 6',
        },
    ],
    [
        'c2f464aa-a291-435d-b465-f35c5466f19c',
        {
            count: 2,
            description: 'Nice Product 7',
            id: 'c2f464aa-a291-435d-b465-f35c5466f19c',
            price: 23,
            title: 'Nice Product 7',
        },
    ],
    [
        '00345b0c-d0fb-4076-836b-b67c715f3a4e',
        {
            count: 3,
            description: 'Nice Product 8',
            id: '00345b0c-d0fb-4076-836b-b67c715f3a4e',
            price: 15,
            title: 'Nice Product 8',
        },
    ],
]);
