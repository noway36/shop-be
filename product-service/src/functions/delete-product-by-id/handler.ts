import 'source-map-support/register';

import { ErrorResponse, formatJSONResponse, logRequestData } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import { deleteProductDataById } from '@libs/data-access';

import { HttpCode } from 'src/constants';
import { APIGatewayProxyHandler } from 'aws-lambda';

export const deleteProductById: APIGatewayProxyHandler = async (event, context) => {
    context.callbackWaitsForEmptyEventLoop = false;
    logRequestData(event, 'add-product');

    try {
        const productId = event.pathParameters?.productId;

        if (productId != null) {
            const deletedProductId: string | undefined = await deleteProductDataById(productId);

            if (deletedProductId != null) {
                return formatJSONResponse({
                    deletedId: deletedProductId,
                });
            }
        }

        return new ErrorResponse(`Failed to delete product with id: ${productId}`, HttpCode.InternalServerError);
    } catch (error) {
        return new ErrorResponse(`Failed to delete product: ${error}`, HttpCode.InternalServerError);
    }
};

export const main = middyfy(deleteProductById);
