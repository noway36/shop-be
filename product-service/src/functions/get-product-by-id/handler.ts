import 'source-map-support/register';

import { APIGatewayProxyHandler } from 'aws-lambda';
import { ErrorResponse, formatJSONResponse, logRequestData } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import { getProductDataById } from '@libs/data-access';

import { HttpCode } from 'src/constants';

export const getProductById: APIGatewayProxyHandler = async (event, context) => {
    context.callbackWaitsForEmptyEventLoop = false;
    logRequestData(event, `get-product-by-id`);

    try {
        const productId = event.pathParameters?.productId;

        if (productId != null) {
            const resultProduct = await getProductDataById(productId);

            if (resultProduct != null) {
                return formatJSONResponse({
                    product: resultProduct,
                });
            }
        }

        return new ErrorResponse(`Product with id: ${productId} not found`, HttpCode.NotFound);
    } catch (error) {
        return new ErrorResponse(`Failed To get Product: ${error}`, HttpCode.InternalServerError);
    }
};

export const main = middyfy(getProductById);
