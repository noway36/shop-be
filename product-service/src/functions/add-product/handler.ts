import 'source-map-support/register';

import {
    ErrorResponse,
    formatJSONResponse,
    logRequestData,
    ValidatedEventAPIGatewayProxyEvent,
} from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import { addProductData } from '@libs/data-access';

import { HttpCode } from 'src/constants';
import { IProduct } from 'src/products-list';
import schema from './schema';

export const addProduct: ValidatedEventAPIGatewayProxyEvent<typeof schema> = async (event, context) => {
    context.callbackWaitsForEmptyEventLoop = false;
    logRequestData(event, 'add-product');

    try {
        const product: Omit<IProduct, 'id'> = event.body?.product;
        if (product != null) {
            const insertedProduct: IProduct | undefined = await addProductData(product);

            if (insertedProduct != null) {
                return formatJSONResponse({
                    product: insertedProduct,
                });
            }
        }

        return new ErrorResponse(`Failed to insert product: ${product}`, HttpCode.InternalServerError);
    } catch (error) {
        return new ErrorResponse(`Failed to insert product: ${error}`, HttpCode.InternalServerError);
    }
};

export const main = middyfy(addProduct);
