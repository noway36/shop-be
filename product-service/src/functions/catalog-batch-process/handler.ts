import 'source-map-support/register';
import { SQSEvent } from 'aws-lambda';

import { formatJSONResponse } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';
import { addProductData } from '@libs/data-access';

import { IProduct } from 'src/products-list';
import { Publisher } from 'src/publisher';

export const catalogBatchProcess = async (event: SQSEvent, context) => {
    context.callbackWaitsForEmptyEventLoop = false;
    console.log(`SQS Event: ${JSON.stringify(event)}`);

    const products: Omit<IProduct, 'id'>[] = event.Records.map(({ body }) => JSON.parse(body));
    const insertedProducts: IProduct[] = [];

    for (const product of products) {
        try {
            const insertedProduct = await addProductData(product);
            if (insertedProduct != null) {
                insertedProducts.push(insertedProduct!);
                console.log(`product: ${JSON.stringify(insertedProduct)} has been uploaded`);

                await Publisher.publishProduct(insertedProduct!);
            }
        } catch (error) {
            console.log(`Error: ${error}`);
        }
    }

    return formatJSONResponse({ products: insertedProducts });
};

export const main = middyfy(catalogBatchProcess);
