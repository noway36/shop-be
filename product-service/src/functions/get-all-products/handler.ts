import 'source-map-support/register';

import { APIGatewayProxyHandler } from 'aws-lambda';
import { ErrorResponse, formatJSONResponse } from '@libs/apiGateway';
import { middyfy } from '@libs/lambda';

import { PRODUCT_BY_ID_MAP } from 'src/products-list';
import { HttpCode } from 'src/constants';

export const getAllProducts: APIGatewayProxyHandler = async () => {
    try {
        const products = Array.from(PRODUCT_BY_ID_MAP.values());

        return formatJSONResponse({
            products,
        });
    } catch (error) {
        return new ErrorResponse(`Failed to get products list`, HttpCode.InternalServerError);
    }
};

export const main = middyfy(getAllProducts);
