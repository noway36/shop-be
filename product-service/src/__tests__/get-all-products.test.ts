import { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import * as mockContext from 'aws-lambda-mock-context';

import { HttpCode } from 'src/constants';
import { IProduct, PRODUCT_BY_ID_MAP } from 'src/products-list';
import { getAllProducts } from '../functions/get-all-products/handler';

describe('Unit test for getAllProducts handler', () => {
    const ctx = mockContext.default();
    const cb = () => null;
    const event = {};

    it('should return array of products', async () => {
        const result = (await getAllProducts(event as APIGatewayProxyEvent, ctx, cb)) as APIGatewayProxyResult;
        const products: IProduct[] = JSON.parse(result.body).products;

        expect(result.statusCode).toEqual(HttpCode.Ok);
        expect(Array.isArray(products)).toEqual(true);
    });

    it('should return proper products', async () => {
        const result = (await getAllProducts(event as APIGatewayProxyEvent, ctx, cb)) as APIGatewayProxyResult;
        const products: IProduct[] = JSON.parse(result.body).products;

        products.forEach(product => {
            expect(product).toMatchSnapshot({ ...PRODUCT_BY_ID_MAP.get(product.id) });
        });
    });
});
