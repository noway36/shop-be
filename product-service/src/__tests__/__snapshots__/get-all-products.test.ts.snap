// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for getAllProducts handler should return proper products 1`] = `
Object {
  "count": 4,
  "description": "Nice Product 1",
  "id": "9a14812c-07e1-463e-9fa1-1a5179f48e61",
  "price": 2,
  "title": "Nice Product 1",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 2`] = `
Object {
  "count": 6,
  "description": "Nice Product 2",
  "id": "164032bb-6512-4d5b-aae8-313fc365f012",
  "price": 10,
  "title": "Nice Product 2",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 3`] = `
Object {
  "count": 7,
  "description": "Nice Product 3",
  "id": "b3595014-634c-4b47-8250-cce3867d4919",
  "price": 23,
  "title": "Nice Product 3",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 4`] = `
Object {
  "count": 12,
  "description": "Nice Product 4",
  "id": "504e5779-c1bc-4498-a792-3b12839b4a7d",
  "price": 15,
  "title": "Nice Product 4",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 5`] = `
Object {
  "count": 7,
  "description": "Nice Product 5",
  "id": "e7a1919b-0bb8-4511-980b-553a332be219",
  "price": 23,
  "title": "Nice Product 5",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 6`] = `
Object {
  "count": 8,
  "description": "Nice Product 6",
  "id": "00e95255-4243-426c-93e7-b9504a8bff89",
  "price": 15,
  "title": "Nice Product 6",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 7`] = `
Object {
  "count": 2,
  "description": "Nice Product 7",
  "id": "c2f464aa-a291-435d-b465-f35c5466f19c",
  "price": 23,
  "title": "Nice Product 7",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 8`] = `
Object {
  "count": 3,
  "description": "Nice Product 8",
  "id": "00345b0c-d0fb-4076-836b-b67c715f3a4e",
  "price": 15,
  "title": "Nice Product 8",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 9`] = `
Object {
  "count": 59,
  "description": "Nice Product 9",
  "id": "8db0532c-6ca2-4abd-8619-e49f84571141",
  "price": 9,
  "title": "Nice Product 9",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 10`] = `
Object {
  "count": 3,
  "description": "Nice Product 10",
  "id": "f0f4cccd-82a4-4fde-8586-a3a0df915fd6",
  "price": 123,
  "title": "Nice Product 10",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 11`] = `
Object {
  "count": 33,
  "description": "Nice Product 11",
  "id": "bf6086a8-0dea-4cec-95a2-e8057cdb6176",
  "price": 5,
  "title": "Nice Product 11",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 12`] = `
Object {
  "count": 1,
  "description": "Nice Product 12",
  "id": "5615ec40-8b88-454b-9389-3a3d8cfaa32c",
  "price": 109,
  "title": "Nice Product 12",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 13`] = `
Object {
  "count": 5,
  "description": "Nice Product 13",
  "id": "26424be3-8749-4db2-a753-d0847c0c41d5",
  "price": 322,
  "title": "Nice Product 13",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 14`] = `
Object {
  "count": 666,
  "description": "Nice Product 14",
  "id": "77d41be9-e998-4961-8c4d-ccf82807f024",
  "price": 666,
  "title": "Nice Product 14",
}
`;

exports[`Unit test for getAllProducts handler should return proper products 15`] = `
Object {
  "count": 6,
  "description": "Nice Product 15",
  "id": "b5942060-a9d2-42ff-8a16-5bf8462a119a",
  "price": 2,
  "title": "Nice Product 15",
}
`;
